h2/src/main/org/h2/mvstore/MVStore.java

```diff
@@ -1982,17 +1985,24 @@ private boolean validateRewrite(Set<Integer> set) {
      * @return the page
      */
     Page readPage(MVMap<?, ?> map, long pos) {
-        if (!DataUtils.isPageSaved(pos)) {
-            throw DataUtils.newIllegalStateException(
-                    DataUtils.ERROR_FILE_CORRUPT, "Position 0");
-        }
-        Page p = cache == null ? null : cache.get(pos);
-        if (p == null) {
-            ByteBuffer buff = readBufferForPage(pos, map.getId());
-            p = Page.read(buff, pos, map);
-            cachePage(p);
+        try {
+            if (!DataUtils.isPageSaved(pos)) {
+                throw DataUtils.newIllegalStateException(
```

Comment:
BTW, where exceptions are thrown during recovery of such databases? In this line, or in the code below it?

Reply:
From what I saw, it's from below, in the Chunk.readBufferForPage().

Reply:
OK. If so, the execution flow in this method is fine, an additional check for `recoveryMode` isn't reasonable.

LGTM.